scripts/core/CUDA.rst

```diff
@@ -0,0 +1,158 @@
+<%
+    OneApi=tags['$OneApi']
+    x=tags['$x']
+    X=x.upper()
+%>
+
+==========================
+CUDA UR Reference Document
+==========================
+
+This document gives general guidelines of how to use UR to load and build
+programs, and execute kernels on a CUDA device.
+
+Device code
+===========
+
+A CUDA device image may be made of PTX and/or SASS, two different kinds of
+device code for NVIDIA GPUs.
+
+CUDA device images can be generated by a CUDA-capable compiler toolchain. Most
+CUDA compiler toolchains are capable of generating PTX, SASS and/or bundles of
+PTX and SASS.
+
+PTX
+---
+
+PTX is a high level NVIDIA ISA which can be JIT compiled at runtime by the CUDA
+driver. In UR, this JIT compilation happens at ${x}ProgramBuild, where PTX is
+assembled into device specific SASS which then can run on device.
+
+PTX is forward compatible, so PTX generated for `.target sm_52` will be JIT
+compiled without issue for devices with a greater compute capability than
+`sm_52`. Whereas PTX generated for `sm_80` cannot be JIT compiled for an
+`sm_60` device.
+
+An advantage of using PTX over SASS is that one code can run on multiple
+devices. However, PTX generated for an older arch may not give access to newer
+hardware instructions, such as new atomic operations, or tensor core
+instructions.
+
+JIT compilation has some overhead at ${x}ProgramBuild, especially if the program
+that is being loaded contains multiple kernels. The `ptxjitcompiler` keeps a
+JIT cache, however, so this overhead is only paid the first time that a program
+is built. JIT caching may be turned off by setting the environment variable
+`CUDA_CACHE_DISABLE=1`.
+
+SASS
+----
+
+SASS is a device specific binary which may be produced by `ptxas` or some other
+tool. SASS is specific to an individual arch and is not portable across arches.
+
+A SASS file may be stored as a `.cubin` file by NVIDIA tools.
+
+UR Programs
+===========
+
+A ${x}_program_handle_t has a one to one mapping with the CUDA driver object
+`CUModule <https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MODULE.html#group__CUDA__MODULE>`__
```

Comment:
Links should only have a single trailing `_`, should the second be a `.` instead?

Reply:
Thanks change made